provisioner/terraform/timings.go

```diff
@@ -0,0 +1,238 @@
+	"github.com/coder/coder/v2/provisionersdk/proto"
+)
+
+type timingKind string
+
+// Copied from https://github.com/hashicorp/terraform/blob/01c0480e77263933b2b086dc8d600a69f80fad2d/internal/command/jsonformat/renderer.go
+// We cannot reference these because they're in an internal package.
+const (
+	timingApplyStart        timingKind = "apply_start"
+	timingApplyProgress     timingKind = "apply_progress"
+	timingApplyComplete     timingKind = "apply_complete"
+	timingApplyErrored      timingKind = "apply_errored"
+	timingProvisionStart    timingKind = "provision_start"
+	timingProvisionProgress timingKind = "provision_progress"
+	timingProvisionComplete timingKind = "provision_complete"
+	timingProvisionErrored  timingKind = "provision_errored"
+	timingRefreshStart      timingKind = "refresh_start"
+	timingRefreshComplete   timingKind = "refresh_complete"
+	// Ignored.
+	timingChangeSummary timingKind = "change_summary"
+	timingDiagnostic    timingKind = "diagnostic"
+	timingPlannedChange timingKind = "planned_change"
+	timingOutputs       timingKind = "outputs"
+	timingResourceDrift timingKind = "resource_drift"
+	timingVersion       timingKind = "version"
+	// These are not part of message_types, but we want to track init/graph timings as well.
+	timingInitStart     timingKind = "init_start"
+	timingInitComplete  timingKind = "init_complete"
+	timingInitErrored   timingKind = "init_errored"
+	timingGraphStart    timingKind = "graph_start"
+	timingGraphComplete timingKind = "graph_complete"
+	timingGraphErrored  timingKind = "graph_errored"
+	// Other terraform log types which we ignore.
+	timingLog        timingKind = "log"
+	timingInitOutput timingKind = "init_output"
+)
+
+type timingAggregator struct {
+	mu sync.Mutex
+
+	stage       database.ProvisionerJobTimingStage
+	stateLookup map[uint64]*timingSpan
+}
+
+type timingSpan struct {
+	kind                       timingKind
+	start, end                 time.Time
+	stage                      database.ProvisionerJobTimingStage
+	action, provider, resource string
+	state                      proto.TimingState
+}
+
+// newTimingAggregator creates a new aggregator which measures the duration of resource init/plan/apply actions; stage
+// represents the stage of provisioning the timings are occurring within.
+func newTimingAggregator(stage database.ProvisionerJobTimingStage) *timingAggregator {
+	return &timingAggregator{
+		stage:       stage,
+		stateLookup: make(map[uint64]*timingSpan),
+	}
+}
+
+// ingest accepts a timing span at a certain timestamp and assigns it a state according to the kind of timing event.
+// We memoize start & completion events, and then calculate their total duration in aggregate.
+// We ignore progress events because we only care about the full duration of the action (delta between *_start and *_complete events).
+func (t *timingAggregator) ingest(ts time.Time, s *timingSpan) {
+	if s == nil {
+		return
+	}
+
+	s.stage = t.stage
```

Comment:
No need for mutex lock while accessing `t` here? Might be good to document on the struct what the mutex protects, e.g.

```go
type timingAggregator struct {
	stage       database.ProvisionerJobTimingStage

	mu sync.Mutex // Protects following.
	stateLookup map[uint64]*timingSpan
}
```

Reply:
Good call. It's only to protect the map right now.